C7/UIElements/RightClickMenu.cs

```diff
@@ -100,45 +101,94 @@ public override void _Input(InputEvent @event)
 		// say that events reach children before their parents, but the catch is that there are three phases of input processing. The "input"
 		// phase, this function, then "gui input", and finally "unhandled input". If a control eats an event during the "input" phase it won't
 		// proceed to the "gui input" phase where buttons actually respond to it.)
-		} else if (! ((@event is InputEventMouse) && mouseOverMenu))
+		} else if (!((@event is InputEventMouse) && mouseOverMenu)) {
 			this.AcceptEvent();
+		}
 	}
 }
 
 public class RightClickTileMenu : RightClickMenu
 {
-	public RightClickTileMenu(Game game, Tile tile) : base(game)
-	{
+	public RightClickTileMenu(Game game, Tile tile) : base(game) {
 		ResetItems(tile);
 	}
 
-	// assumeUnfortified is a unit that will be considered unfortified regardless of what its isFortified field holds. It's used to work around
-	// the fact that unfortification is done asynchronously through a message to the engine.
-	public void ResetItems(Tile tile, MapUnit assumeUnfortified = null)
-	{
+	// uiUpdatedUnitStates maps unit guid to a boolean that is true if they were fortified
+	// and false if they were selected in the previous action. This is to update the UI
+	// since the actions update the engine asynchronously and otherwise the UI may not
+	// reflect these changes immediately.
+	public void ResetItems(Tile tile, Dictionary<string, bool> uiUpdatedUnitStates = null) {
 		RemoveAll();
+
+		bool isUnitFortified(MapUnit unit) {
+			if (uiUpdatedUnitStates is null || !uiUpdatedUnitStates.ContainsKey(unit.guid)) {
+				return unit.isFortified;
+			}
+			return uiUpdatedUnitStates[unit.guid];
+		}
+
+		string getUnitAction(MapUnit unit, bool isFortified) {
```

Comment:
I'd kind of favor making this and isUnitFortified private methods that aren't embedded in another private method.  That would keep the focus of this method at a higher level.

Reply:
sounds good, I'll change that